Improve: parse doc comments with Odoc

Julow opened the pull request:
(issue #571)

Work in progress, but the current state is relatively clean and can be reviewed.

Rewrite `Fmt_odoc` and `Normalize` to use Odoc.
It adds a lot of regressions and fixes a few bugs.

Odoc `1.4.0` exports its internal libraries, it's no longer needed to vendor it. Other small problems remain.

jberdine commented on src/Sugar.mli:
```diff
@@ -97,9 +97,9 @@ val ite :
-        else e3
+if c1 then e1
+else if c2 then e2
+else e3
```
Is this indentation intended?

Julow replied:
A bit of work is needed to fix this. It is intended for verbatim blocks to avoid pushing ascii tables to the margin.

jberdine replied:
Ok. I don't really know what the best choice ought to be, but expect that people will want to be able to keep things out of the left margin. But maybe not  if that means having inexplicable extra space in the generated docs.

Julow replied:
Whitespaces are preserved in verbatim/code blocks. The user is free to indent it.

jberdine commented on src/dune:
```diff
@@ -21,7 +21,7 @@ Jbuild_plugin.V1.send
  (flags (:standard -open Import))
  %s
- (libraries cmdliner format_ fpath import ocaml-migrate-parsetree octavius unix uutf))
+ (libraries cmdliner format_ fpath import ocaml-migrate-parsetree unix uutf odoc.model odoc.parser))
```
alphabetize

jberdine commented on test/passing/doc_comments.mli:
```diff
@@ -13,23 +13,25 @@ val escape : string -> string
 type title = string * int * string * string * string
 
-(** {1 Standard section names} *)
+(** {1:standard-section-names Standard section names} *)
```
Is it necessary to add `:standard-section-names`? Seems very redundant.

Julow replied:
This is not optional in Odoc's ast. Work is needed to remove them.

jberdine replied:
Ok I see. Are they generated if omitted, so that we could detect if they have the generated name and then not print them?

Julow replied:
Users may want to explicitly add the label. Reading the original string is possible but tricky. A good solution would be to modify Odoc.

jberdine commented on test/passing/doc_comments.mli:
```diff
@@ -62,24 +66,37 @@ val doc_to_plain :
 
 val k : k
 (** this is a comment
+
```
Are these extra open lines intended / necessary?

Julow replied:
Not intended. I added the blank line to separate lists and paragraphs. I'll add a special case for tags (and module references, verbatim and code blocks).

hhugo replied:
There are sill new empty lines

hhugo replied:
ignore me. I didn't see the the reference file.

jberdine commented on test/passing/doc_comments.mli:
```diff
@@ -180,7 +201,7 @@ val x : x
 (** [a]'c [b]'s [c]'c *)
 
-(** return true if {%\gamma(lhs) \subseteq \gamma(rhs)%} *)
+(** return true if [\gamma(lhs) \subseteq \gamma(rhs)] *)
```
The `{%` `%}` markup ought to be preserved, right?

Julow replied:
In this case (and the next one) Odoc gives a warning and we could refuse to format with an error.

jberdine replied:
Oh. So odoc doesn't support the whole ocamldoc language? `{%` and `%}` are documented markup.

Julow replied:
Odoc gives a warning and turn it into a code span:
```
'{%...%}' (raw markup) needs a target language.
Suggestion: try '{%html:...%}'.
```
Warnings are easily handled, I did not code this part yet.
In case of warning, do you think we should print the warning and let the doc comment as is or refuse to format the whole file ?
The first seems more reasonable but without a `-Werror` flag (or maybe when `--check` is set), it could stay unnoticed.

jberdine replied:
This seems like an odoc bug / divergence from spec to me. The `latex` language is the default: http://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#sss%3Atarget-specific-syntax

Julow replied:
This is now fixed.

jberdine commented on test/passing/doc_comments.mli:
```diff
@@ -189,7 +210,7 @@ val x : x
 (** for [Global]s *)
 
-(** generic command: ∀xs.{foot}-{post} *)
+(** generic command: ∀xs.[foot]-[post] *)
```
We need to avoid rewrites like this, they are semantically meaningful. If the original is ill-formed somehow, better to fail.

Julow replied:
It is:
```
'{abc}': bad markup.
Suggestion: did you mean '{!abc}' or '[abc]'?
```

jberdine replied:
Ok, thanks. I think it would be better to fail in such cases. Or at least warn, or have an option to choose. There are, I expect, lots of docstrings in the wild that are ill-formed, and it isn't only ocamlformat that will be doing the wrong thing with them. So probably better to give users feedback as early as possible. Hopefully we can get line numbers, etc. for an error message.

hhugo commented:
How far are we to have this PR ready ? I'm wondering if we should try to fix Octavious for 4.08 or if this PR will land in reasonable time ?

Julow commented:
Sorry for the delay.

I think this PR can land in reasonable time from now. I'm not 100% confident it can before the OCaml release.

It is not ready for merging but should be ready for a new review.

I rebased on master.

Since last reviews:
- Removed unecessary blank lines between some elements
- Don't format comments when Odoc generated warnings
	Because Odoc may have lost informations. In practice Odoc try its best to not to.
- Use Odoc parser's `Ast` instead of `Model.Comment`
	It is a bit simpler and the parser is a lot simpler to call.
	I made a PR to Odoc: https://github.com/ocaml/odoc/pull/351
	This fixes label on headings
- Removed terminating break in list elements with heavy syntax
- Fixed the formatting of references
	This part is ridiculously hard. Odoc could be modified to improve this but it would be a big change.

Julow commented on src/Fmt_odoc.ml:
```diff
@@ -9,120 +9,218 @@
+(** Escape characters if they are not already escaped. [escapeworthy] should
+    return [true] if the character should be escaped, [false] otherwise. *)
+let ensure_escape ?(escape_char = '\\') ~escapeworthy s =
+  let module E = String.Escaping in
```
It's local but I agree. It is used only once anyway.

Julow commented on src/Fmt_odoc.ml:
```diff
@@ -9,120 +9,218 @@
+
+let fmt_if_not_empty lst fmt = fmt_if (not (List.is_empty lst)) fmt
+
+let ign_loc f with_loc = f with_loc.Location_.value
```
Changed.

Julow commented on src/Fmt_odoc.ml:
```diff
@@ -9,120 +9,218 @@
+  | `Word w ->
+      (* Escape lines starting with '+' or '-' *)
+      let escape =
+        if String.length w > 0 && Char.(w.[0] = '+' || w.[0] = '-') then
```
Yes.

Julow commented on src/Fmt_odoc.ml:
```diff
@@ -152,39 +250,85 @@ and fmt_text txt =
+  | `Paragraph elems -> hovbox 0 (fmt_inline_elements elems)
+  | `Code_block s ->
+      vbox 0 (fmt "{[@;<1 -999>" $ str_verbatim s $ fmt "@ ]}")
+  | `Verbatim s -> vbox 0 (fmt "{v@;<1 -999>" $ str_verbatim s $ fmt "@ v}")
```
The goal is to remove any indentation. What are the alternatives ?

Julow replied:
I changed the code.

Julow commented:
Now that my PRs on Odoc have been merged (https://github.com/ocaml/odoc/pull/351, https://github.com/ocaml/odoc/pull/364), this can be implemented more easily.

Rebased on master and fixed a few bugs to to the rebase and changes in Odoc.
Currently, it is needed to pin Odoc to master.

avsm commented on ocamlformat.opam:
```diff
@@ -29,7 +29,7 @@ depends: [
   "fpath"
   "ocaml-migrate-parsetree" {>= "1.3.1"}
-  "octavius" {>= "1.2.0"}
+  "odoc" {>= "1.4.0"}
```
```suggestion
  "odoc" {>= "1.4.1"}
```

just merged https://github.com/ocaml/opam-repository/pull/14349

Julow replied:
Finally available to `opam update`. Upgraded.

avsm commented on ocamlformat_reason.opam:
```diff
@@ -28,7 +28,7 @@ depends: [
   "fpath"
   "ocaml-migrate-parsetree" {>= "1.0.10"}
-  "octavius" {>= "1.2.0"}
+  "odoc" {>= "1.4.0"}
```
```suggestion
  "odoc" {>= "1.4.1"}
```

Julow commented:
Rebased on master. The history is a mess, this should be reviewed file by file.

gpetiot commented:
The diff with test_branch is:
```
diff --git a/src/Sugar.mli b/src/Sugar.mli
index 81514eb..db5f9f6 100644
--- a/src/Sugar.mli
+++ b/src/Sugar.mli
@@ -93,9 +93,9 @@ val ite :
     if-then-else expressions, e.g.:
 
     {[
-        if c1 then e1
-        else if c2 then e2
-        else e3
+if c1 then e1
+else if c2 then e2
+else e3
     ]}
 
     will return the following list:
@@ -141,11 +141,9 @@ val polynewtype :
     pattern [pat] with body [exp]. e.g.:
 
     {[
-        let f: 'r 's. 'r 's t = fun (type r) -> fun (type s) -> (e : r s t)
+let f: 'r 's. 'r 's t = fun (type r) -> fun (type s) -> (e : r s t)
     ]}
 
     Can be rewritten as:
 
-    {[
-        let f: type r s. r s t = e
-    ]} *)
+    {[ let f: type r s. r s t = e ]} *)
```

I still think starting the verbatim code blocks at column 0 looks weird, but it shouldn't delay the merge.

gpetiot commented on src/Fmt_ast.ml:
```diff
@@ -415,14 +415,19 @@ let fmt_label lbl sep =
+        List.iter warnings ~f:(fun w ->
+            let msg = Odoc_model.Error.to_string w in
+            Caml.Format.eprintf
+              "%a:@,Warning: Invalid documentation comment:@,%s\n%!"
```
Maybe this warning should be printed only when `parse-docstrings=true`

Julow commented:
Thanks for the review. I improved the indentation of code blocks. I keep the behavior of fitting single line blocks.
I cleaned up a little and moved the printing of the warning.

What is not yet implemented is:
- Preserve list syntax
- Preserve whitespaces

I'm not sure if we should and how to preserve whitespaces. This should be done later on its own PR.

gpetiot commented:
`tools/test_branch.sh odoc_instead_of_octavius 'parse-docstrings=true' 'parse-docstrings=true'` shows the formatting of some docstrings became unstable (probably due to 87375aa)

Julow commented:
The unstable docstring errors are from the `master` build. They should all be fixed by this PR. This commit is clean.

gpetiot commented:
> The unstable docstring errors are from the `master` build. They should all be fixed by this PR. This commit is clean.

For example `ocamlformat test/code/infer/infer/src/biabduction/Prop.mli --no-version-check --parse-docstrings` fails with master but also with this PR, so it's not really a regression, but it would be nice if it would be fixed.

But we can fix it in another PR, feel free to merge.

Julow commented:
I can't reproduce this bug. It fails on master but not on this branch. Do you have an older version of this branch ? (I rebased it 3 days ago)

gpetiot commented:
> I can't reproduce this bug. It fails on master but not on this branch. Do you have an older version of this branch ? (I rebased it 3 days ago)

No I use your rebased version.

hhugo commented on test/passing/doc_comments.mli:
```diff
@@ -123,9 +140,9 @@ val x : x
-      let verbatim s =
-        s |> String.split_lines |> List.map ~f:String.strip
-        |> fun s -> list s "@," Fmt.str
+let verbatim s =
```
Why is still code block indented to the left ?

hhugo replied:
ignore me. I didn't see the the reference file.

hhugo commented on test/passing/js_source.ml.ref:
```diff
@@ -9759,7 +9759,8 @@ if fffffffffffffff aaaaa bb
 else aaaaaaaaaaaa qqqqqqqqqqq
 
-include Base.Fn (** @open *)
+(** @open *)
```
Why did the comment move before ?

hhugo commented on test/passing/tag_only.ml:
```diff
@@ -102,26 +103,33 @@ class b =
 
-open Module  (** @deprecated  *)
+open Module
+(** @deprecated  *)
```
Why do we have a line break here ? I though `doc-comments-tag-only=fit` was for keeping theses as one-liner ?

Julow replied:
Must have been lost in a rebase. Fixed.

Julow commented:
Fixed printing of `@see` tags and tag-only comments and removed unecessary spaces.

Julow commented:
I tested this with other projects and found some bugs:

- `*` character in the left margin not removed and treated as text

```diff
-val qubesdb: qubesdb typ
-(** For the Qubes target, the Qubes database from which to look up
- *  dynamic runtime configuration information. *)
+val qubesdb : qubesdb typ
+(** For the Qubes target, the Qubes database from which to look up * dynamic
+    runtime configuration information. *)
```

- Balanced brackets do not need to be escaped

```diff
-(** [intersperse xs ~sep] places [sep] between adjacent elements of [xs].  For example,
-    [intersperse [1;2;3] ~sep:0 = [1;0;2;0;3]]. *)
+(** [intersperse xs ~sep] places [sep] between adjacent elements of [xs]. For example,
+    [intersperse \[1;2;3\] ~sep:0 = \[1;0;2;0;3\]]. *)
```

- Breaks in inline code spans are preserved

```diff
-(** [drop l n] returns [l] without the first [n] elements, or the empty list if [n >
-    length l].  [drop l n] is equivalent to [snd (split_n l n)]. *)
+(** [drop l n] returns [l] without the first [n] elements, or the empty list if
+    [n >
+    length l]. [drop l n] is equivalent to [snd (split_n l n)]. *)
```

- Indentation of code blocks is not working when text starts on the same line as the opening

```diff
 (** Abc
-    - {[ a
-         b
+    - {[
+        a
+                b
       ]} *)
```

They may not be bugs or may be fixed later. I'm looking at the last one.

Julow commented on test/passing/doc_comments.mli.ref:
```diff
@@ -0,0 +1,331 @@
+    {[
+      First line
+            on the same line
+            as opening
```
This requires a fix to Odoc. https://github.com/ocaml/odoc/pull/370

Julow replied:
It is now merged but will be fixed in OCamlformat only after the next release of Odoc.

Julow commented on test/passing/doc_comments.mli.ref:
```diff
@@ -330,9 +330,9 @@ end
 
-    [Very looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong] *)
+    [Very
+    looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong] *)
```
Whitespaces are now normalized and allowed to break in code spans.

gpetiot commented on src/Fmt_odoc.ml:
```diff
@@ -145,7 +145,7 @@ let rec fmt_inline_element : inline_element -> Fmt.t = function
   | `Code_span s ->
       let s = escape_brackets s in
-      hovbox 0 (wrap "[" "]" (str s))
+      hovbox 0 (wrap "[" "]" (str_normalized s))
```
The CI is fixed if you use `~escape:false`, is it necessary to escape all these characters?

Julow replied:
It's not necessary to escape all of them, only `[` and `]`. Also `str_normalized` will remove consecutive whitespaces, which may not be what we want.

Julow replied:
I changed to only escape brackets and to keep consecutive whitespaces. The CI is fixed.

gpetiot replied:
I'm not convinced regarding keeping consecutive whitespaces, I think if someone would want to keep the spaces he would use the verbatim blocks instead.

Julow replied:
I reverted that part.
